components/logins/src/error.rs

```diff
@@ -97,192 +85,181 @@ pub enum InvalidLogin {
     // is no longer valid. See [InvalidLoginReason] for a list of reasons
     // a record may be considered invalid
-    #[error("{0}")]
-    InvalidRecord(String, InvalidLoginReason),
+    #[error("Invalid login: {0}")]
+    InvalidRecord(InvalidLogin),
 
-    /// This error is emitted when migrating from a sqlcipher database in two cases:
-    /// 1. An incorrect key is used to to open the login database
-    /// 2. The file at the path specified is not a sqlite database.
-    /// NOTE: Dropping sqlcipher means we will drop (1), so should rename it
-    #[error("InvalidKey error: {0}")]
-    InvalidKey(String),
+    /// This is thrown if `update()` is performed with a record whose ID
+    /// does not exist.
+    #[error("No record with guid exists (when one was required): {0}")]
+    NoSuchRecord(String),
 
     /// Error encrypting/decrypting logins data
-    #[error("Crypto Error: {0}")]
+    #[error("Encryption error: {0}")]
     CryptoError(String),
 
+    /// This indicates that the sync authentication is invalid, likely due to having
+    /// expired.
+    #[error("SyncAuthInvalid error: {0}")]
+    SyncAuthInvalid(String),
+
     /// This error is emitted if a request to a sync server failed.
-    /// We can probably kill this? The sync manager is what cares about this.
+    ///
+    /// Once iOS is using the sync manager, we can probably kill this.  Since the sync manager will
+    /// then be handling the error.
     #[error("RequestFailed error: {0}")]
     RequestFailed(String),
 
-    /// This error is emitted if a sync or other operation is interrupted.
-    #[error("Interrupted error: {0}")]
+    /// Operation was interrupted by the user
+    #[error("Operation interrupted: {0}")]
     Interrupted(String),
+
+    /// Catch-all for all other errors
+    #[error("Unexpected error: {0}")]
+    UnexpectedLoginsError(String),
 }
 
-/**
- * A reason a login may be invalid
- */
-#[derive(Debug)]
-pub enum InvalidLoginReason {
-    /// Origins may not be empty
-    EmptyOrigin,
-    /// Passwords may not be empty
-    EmptyPassword,
-    /// The login already exists
-    DuplicateLogin,
-    /// Both `httpRealm` and `formActionOrigin` are non-null
-    BothTargets,
-    /// Both `httpRealm` and `formActionOrigin` are null
-    NoTarget,
-    /// Login has illegal field
-    IllegalFieldValue,
+impl From<LoginsError> for LoginsStorageError {
+    fn from(error: LoginsError) -> LoginsStorageError {
+        // We convert errors before sending them across the API boundary to the consuming
+        // application, so this is a good time to report them.
+        error.report();
+
+        match error {
+            LoginsError::InvalidLogin(inner) => Self::InvalidRecord(inner),
+            LoginsError::NoSuchRecord(guid) => Self::NoSuchRecord(guid),
+            LoginsError::CryptoError(inner) => Self::CryptoError(inner.to_string()),
+            LoginsError::InvalidKey => Self::CryptoError("InvalidKey".to_string()),
```

Comment:
This one is slightly risky, since the recommendation for `CryptoError` is to delete your database and start over.  I'm pretty sure that the same thing is true for `InvalidKey` since:
  - An invalid local encryption key is effectively the same as a valid key that can't decrypt database data
  - We basically don't return `InvalidKey` for sqlcipher databases anymore.
  - We could maybe return `InvalidKey` is during the sqlcipher migration, but that also means indicates the app should delete the database and start fresh.

That's my logic at least, does that make sense to you?  @mhammond @lougeniaC64

Reply:
@bendk This makes sense to me. At present we recreate the database and/or report the error with both `CryptoError` and `InvalidKey`. We can always update this code if the handling needs to diverge in the future (although I can't imagine why that would happen).